# 10.15 Exercises
You can download solutions to these exercises from http://thinkpython2.com/code/list_exercises.py.

**Exercise 10.1.** *Write a function called nested_sum that takes a list of lists of integers and adds up the elements from all of the nested lists. For example:*

In [1]:
def nested_sum(t):
    s=0
    for line in t:
        s+=sum(line)
    return s

t=[[1, 2], [3], [4, 5, 6]]
nested_sum(t)

21

**Exercise 10.2.** *Write a function called* `cumsum` *that takes a list of numbers and returns the cumulative sum; that is, a new list where the ith element is the sum of the first i + 1 elements from the original list. For example:*

In [2]:
def cumsum(t):
    new=[]
    i=0
    while i < len(t):
        i+=1
        new += [sum(t[:i])]
    return new   
    
        
t = [1, 2, 3]
cumsum(t)

[1, 3, 6]

**Exercise 10.3.** *Write a function* called `middle` *that takes a list and returns a new list that contains all but the first and last elements. For example:*

In [3]:
def middle(t):
    new=t[1:(len(t)-1)]
    return new

t = [1, 2, 3, 4]
middle(t)

[2, 3]

**Exercise 10.4.** *Write a function called* `chop` *that takes a list, modifies it by removing the first and last elements, and returns `None.` For example:*

In [4]:
def chop(t):
    del (t[0],t[len(t)-1])
    return None

t = [1, 2, 3, 4]

chop(t)
print(t)

[2, 3]


**Exercise 10.5.** *Write a function called* `is_sorted` *that takes a list as a parameter and returns `True` if the list is sorted in ascending order and False otherwise. For example:*

In [5]:
def is_sorted(t):
    i=0
    while i < len(t)-1:
        if t[i] <= t[i+1]:
            i += 1
        else:
            return False
    return True

print(is_sorted([1, 2, 2]))
print(is_sorted(['b', 'a']))

True
False


**Exercise 10.6.** *Two words are anagrams if you can rearrange the letters from one to spell the other. Write a function called* `is_anagram` *that takes two strings and returns True if they are anagrams.*

In [6]:
def is_anagram(s1, s2):
    i=0
    s01=s1.replace(' ', '')
    s02=s2.replace(' ', '')

    if len(s01)!=len(s02):
        return False
    elif s01 != s02[::-1]:
        return False
    else:
        return True


s1= 'ana banana'
s2= 'ana canana'
print(is_anagram(s1, s2))
s3= 'subi no onibus'
s4= 'subi no onibus'
print(is_anagram(s3, s4))

False
True


**Exercise 10.7.** *Write a function called* `has_duplicates` *that takes a list and returns True if there is any element that appears more than once. It should not modify the original list.*

In [7]:
def has_duplicates(t):
    for i in range(len(t)):
        for j in range(len(t)):
            if i==j:
                continue
            elif t[i]==t[j]:
                return True
    return False

t=[1,2,3,4]
print(has_duplicates(t))
t=[1,2,3,2]
print(has_duplicates(t))

False
True


**Exercise 10.8.** *This exercise pertains to the so-called Birthday Paradox, which you can read about
at http://en.wikipedia.org/wiki/Birthday_paradox.*

*If there are 23 students in your class, what are the chances that two of you have the same birthday?
You can estimate this probability by generating random samples of 23 birthdays and checking for
matches.*

*Hint: you can generate random birthdays with the randint function in the random module.
You can download my solution from http://thinkpython2.com/code/birthday.py.*

In [8]:
from random import randint

def random_bdays(n):
    """Returns a ramdom list of birthdays
    """
    bd=[]
    
    for i in range(n):
        bd+=[randint(1,365)]
    return bd

def has_duplicates(t):
    for i in range(len(t)):
        for j in range(len(t)):
            if i==j:
                continue
            elif t[i]==t[j]:
                return True
    return False

def counting_matches(n_simul, n):
    matches=0
    for i in range(n_simul):
        if has_duplicates(random_bdays(n)):
            matches+=1
    return matches


n=23
n_simul=1000
    
print("""After {} simulations, 
with {} students each, 
we obtained {} matches.""".format(n_simul, n, counting_matches(n_simul, n)))


After 1000 simulations, 
with 23 students each, 
we obtained 517 matches.


**Exercise 10.9.** *Write a function that reads the file words.txt and builds a list with one element
per word. Write two versions of this function, one using the append method and the other using
the idiom* `t = t + [x]`. *Which one takes longer to run? Why?*

Solution: http://thinkpython2.com/code/wordlist.py .

In [9]:
import time

def make_word_list1():
    """Reads lines from a file and builds a list using append."""
    words = []
    fin = open('words.txt')
    for line in fin:
        word = line.strip()
        words.append(word)
    return words

def make_word_list2():
    """Reads lines from a file and builds a list using +"""
    words = []
    fin = open('words.txt')
    for line in fin:
        word = line.strip()
        words= words+[word]
    return words

start1= time.time()
words1=make_word_list1()
elapsed1=time.time() - start1

print('The first version takes {} seconds.'.format(elapsed1))

start2=time.time()
words2=make_word_list2()
elapsed2=time.time() - start2

print('The second version takes {} seconds.'.format(elapsed2))



def make_word_list3():
    """Reads lines from a file and builds a list using +=
    (This was not required at the exercise)"""
    words = []
    fin = open('words.txt')
    for line in fin:
        word = line.strip()
        words += [word]
    return words

start3=time.time()
words3=make_word_list3()
elapsed3=time.time() - start3

print('The third version takes {} seconds.'.format(elapsed2))

print('The lists have {} words.'.format(len(words1)))

The first version takes 0.6109590530395508 seconds.
The second version takes 237.97663617134094 seconds.
The third version takes 237.97663617134094 seconds.
The lists have 113809 words.


**Exercise 10.10.** *To check whether a word is in the word list, you could use the* `in` *operator, but it would be slow because it searches through the words in order*.

*Because the words are in alphabetical order, we can speed things up with a bisection search (also
known as binary search), which is similar to what you do when you look a word up in the dictionary.
You start in the middle and check to see whether the word you are looking for comes before the word
in the middle of the list. If so, you search the first half of the list the same way. Otherwise you search the second half.*

*Either way, you cut the remaining search space in half. If the word list has 113,809 words, it will
take about 17 steps to find the word or conclude that it’s not there.*

*Write a function called* `in_bisect` *that takes a sorted list and a target value and returns* `True` *if the word is in the list and* `False` *if it’s not.*

*Or you could read the documentation of the* `bisect` *module and use that! Solution: http://thinkpython2.com/code/inlist.py .*

In [2]:
def make_word_list():
    """Reads lines from a file and builds a list using append."""
    words = []
    fin = open('words.txt')
    for line in fin:
        word = line.strip()
        words.append(word)
    return words



aa in list True
alien in list True
allen in list False
zymurgy in list True


In [11]:
"""This module contains a code example related to

Think Python, 2nd Edition
by Allen Downey
http://thinkpython2.com

Copyright 2015 Allen Downey

License: http://creativecommons.org/licenses/by/4.0/
"""

from __future__ import print_function, division

import bisect


def make_word_list():
    """Reads lines from a file and builds a list using append.

    returns: list of strings
    """
    word_list = []
    fin = open('words.txt')
    for line in fin:
        word = line.strip()
        word_list.append(word)
    return word_list


def in_bisect(word_list, word):
    """Checks whether a word is in a list using bisection search.

    Precondition: the words in the list are sorted

    word_list: list of strings
    word: string

    returns: True if the word is in the list; False otherwise
    """
    if len(word_list) == 0:
        return False

    i = len(word_list) // 2
    if word_list[i] == word:
        return True

    if word_list[i] > word:
        # search the first half
        return in_bisect(word_list[:i], word)
    else:
        # search the second half
        return in_bisect(word_list[i+1:], word)


def in_bisect_cheat(word_list, word):
    """Checks whether a word is in a list using bisection search.

    Precondition: the words in the list are sorted

    word_list: list of strings
    word: string
    """
    i = bisect.bisect_left(word_list, word)
    if i == len(word_list):
        return False

    return word_list[i] == word


if __name__ == '__main__':
    word_list = make_word_list()
    
    for word in ['aa', 'alien', 'allen', 'zymurgy']:
        print(word, 'in list', in_bisect(word_list, word))

    for word in ['aa', 'alien', 'allen', 'zymurgy']:
        print(word, 'in list', in_bisect_cheat(word_list, word))


aa in list True
alien in list True
allen in list False
zymurgy in list True
aa in list True
alien in list True
allen in list False
zymurgy in list True


**Exercise 10.11.** *Two words are a “reverse pair” if each is the reverse of the other. Write a program that finds all the reverse pairs in the word list. Solution:* http://thinkpython2.com/code/reverse_pair.py .

**Exercise 10.12.** *Two words “interlock” if taking alternating letters from each forms a new
word. For example, “shoe” and “cold” interlock to form “schooled”. Solution:* http://thinkpython2.com/code/interlock.py. 

Credit: This exercise is inspired by an example at http://puzzlers.org .

1. Write a program that finds all pairs of words that interlock. Hint: don’t enumerate all pairs!
2. Can you find any words that are three-way interlocked; that is, every third letter forms a word, starting from the first, second or third?